adobe/spectrum-web-components
# feat(breadcrumbs): add Breadcrumbs component

## Description

Breadcrumbs show hierarchy and navigational context for a user’s location within an app.
[Spectrum CSS documentation](https://opensource.adobe.com/spectrum-css/breadcrumb.html)
[Spectrum design](https://spectrum.adobe.com/page/breadcrumbs/)

![308627962-4afefbf3-5971-484d-b85b-c62ade9bbf37](https://github.com/user-attachments/assets/8f52e3b8-6fb5-4d8f-83ea-0d2605320529)

This PR adds the following features to the `sp-breadcrumbs` component:

- [x]  Responsive expand / collapse based on available space
- [x]  Responsive expand / collapse based on `sp-breadcrumb-item` addition / removal
- [x] `show-root` attribute to always display the first breadcrumb
- [x] `href` support for each `sp-breadcrumb-item`, which will redirect to provided `href` on click
- [x] `change` event for each `sp-breadcrumb-item` when no `href` is provided, to allow the consumer to perform whatever action needed
- [x] Customisation of the action-menu icon, if the consumer needs it
- [x] Accessibility aspects

Remaining to do:

- [x] `compact` support
- [ ] `multiline` support
- [x] `disabled` support
- [ ] `dragged` support
- [x] Enhance documentation
- [x] Unit tests for `sp-breadcrumb-item` component
- [ ] Truncation of the current route when the name is too long
- [x] There is a bug in Spectrum CSS regarding `large` scale for this component. Elements are a bit unaligned. This is a known issue and it is being investigated. This will probably need a bump in spectrum-css dependency.

## Related issue(s)



- https://github.com/adobe/spectrum-web-components/discussions/4100

## Motivation and context



## How has this been tested?




-   [ ] _General UI layout_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--default)
    2. Compare with [Spectrum CSS](https://opensource.adobe.com/spectrum-css/breadcrumb.html)
-   [ ] _Responsiveness to nr of children_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--default)
    2. Using the Storybook controls Add-on, change the `nrOfItems` parameter. Observe how the Breadcrumbs component reacts to this change, elements are collapsed / expanded based on `max-visible-items` (which is not reactive, its initial value is the one that is persistent)
-   [ ] _Responsiveness to available space_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--resizable-behavior)
    2. Use the resizable container to modify the available space that the Breadcrumbs component has for rendering
    3. Observe how more items are added / removed from the screen
    4. Observe how you can not remove all of them, at least one breadcrumb item should be visible at all times.
 -   [ ] _`show-root`_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--show-root)
    2. Compare with [Spectrum CSS](https://opensource.adobe.com/spectrum-css/breadcrumb.html)
 -   [ ] _Click events_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--default)
    2. Click on any breadcrumb item
    5. Observe the Actions addon
    6. Observe that the last breadcrumb item does not react to click events
 -   [ ] _href support_
    1. Go [here](https://rmanole-breadcrumbs--spectrum-web-components.netlify.app/storybook/?path=/story/breadcrumbs--links)
    2. Click on any breadcrumb item
    4. Observe that the URL of the browser changes to the respective value of the `href` attribute
    5. Observe that the last breadcrumb item does not redirect you anywhere



## Screenshots (if appropriate)

## Types of changes




## Checklist




## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.
